smatch/matcher.py

```diff
@@ -0,0 +1,213 @@
+            dimension `k` added to the end.  If `k=1` then this last
+            dimension is squeezed out.
+        i1 : array-like, int
+            Array of indices for matcher lon/lat.
+            Returned if return_indices is True.
+        i2 : array-like, int
+            Array of indices for query lon/lat.
+            Returned if return_indices is True.
+        d : array-like, float
+            Array of distances in degrees. Same shape as input array with axis
+            of dimension `k` added to the end. If `k=1`, then this last dimension
+            is squeezed out.
+        """
+        if distance_upper_bound is not None:
+            maxd = 2*np.sin(np.deg2rad(distance_upper_bound)/2.)
+        else:
+            maxd = np.inf
+
+        if k != 1 and return_indices:
+            raise NotImplementedError("Indices are only returned for 1-1 matches")
+
+        coords = hp.rotator.dir2vec(lon, lat, lonlat=True).T
+        d, idx = self.tree.query(coords, k=k, p=2, distance_upper_bound=maxd, eps=eps)
+
+        if return_distances:
+            d /= 2
+            np.arcsin(d, out=d, where=np.isfinite(d))
+            d = np.rad2deg(2*d)
+
+        if return_indices:
+            i2, = np.where(np.isfinite(d))
+            i1 = idx[i2]
+            return idx, i1, i2, d
+        else:
+            if return_distances:
+                return idx, d
+            else:
+                return idx
+
+    def query_radius(self, lon, lat, radius, eps=0.0, return_indices=False):
+        """Find all points in (lon, lat) that are within `radius` of the
+        points in the matcher.
+
+        Parameters
+        ----------
+        lon : array-like
+            The longitude in degrees.
+        lat : array-like
+            The latitude in degrees.
+        radius : `float`
+            The match radius in degrees.
+        eps : `float`, optional
+            If non-zero, the set of returned points are correct to within a
+            fraction precision of `eps` being closer than `radius`.
+        return_indices : `bool`, optional
+            Return tuple of (idx, i1, i2, distance) instead of just idx.
+
+        Returns
+        -------
+        idx : `list` [`list` [`int`]]
+            Each row in idx corresponds to each position in matcher lon/lat.
+            The indices in the row correspond to the indices in query lon/lat.
+        i1 : array-like, int
+            Array of indices for matcher lon/lat.
+            Returned if return_indices is True.
+        i2 : array-like, int
+            Array of indices for query lon/lat.
+            Returned if return_indices is True.
+        distance : array-like, float
+            Array of distance (degrees) for each match pair.
+            Returned if return_indices is True.
+        """
+        coords = hp.rotator.dir2vec(lon, lat, lonlat=True).T
+        # The second tree in the match does not need to be balanced, and
```

Comment:
I think this comment should probably be moved up top ... neither tree needs to be balanced as it turns out.

Reply:
done.